File: ch7/circles.ml
open Printf

type 'a point = { x: 'a; y: 'a }
let point (x, y) = { x; y }

type 'a circle = { center: 'a point; radius: 'a } 

let pi = Float.pi
let sqr x = x *. x
let deg_to_rad d = d *. pi /. 180.

(* Returns: 0 inside / 1 border / 2 outside *)
let inside_circle {x; y} { center; radius } =
  let dx, dy = x - center.x, y - center.y in
  compare (dx * dx + dy * dy) (radius * radius) + 1

(** Returns the center of a circle with radius r and goes through 2 points.
    To get the other center, reverse p1 and p2 *)
let circle_2pts_rad p1 p2 r =
  let d2 = sqr (p1.x -. p2.x) +. sqr (p1.y -. p2.y) in
  let det = sqr r /. d2 -. 0.25 in
  if det < 0. then None else
  let h = sqrt det in
  let cx = (p1.x +. p2.x) *. 0.5 +. (p1.y -. p2.y) *. h in
  let cy = (p1.y +. p2.y) *. 0.5 +. (p1.x -. p2.x) *. h in
  Some (point(cx, cy))

let () =
  let circle = { center = point(2, 2); radius = 7 } in
  let inside = point(8, 2) in printf "%d\n" (inside_circle inside circle);
  let border = point(9, 2) in printf "%d\n" (inside_circle border circle);
  let outside = point(10, 2) in printf "%d\n" (inside_circle outside circle);

  let r = float_of_int circle.radius in
  let d = r *. 2. in
  let c = d *. pi in
  let area = pi *. sqr r in
  printf "Diameter = %.2f\n" d;
  printf "Circumference (Perimeter) = %.2f\n" c;
  printf "Area of circle = %.2f\n" area;

  printf "Length of arc   (central angle = 60 degrees) = %.2f\n"
    (60.0 /. 360.0 *. c);
  printf "Length of chord (central angle = 60 degrees) = %.2f\n"
    (sqrt (2. *. sqr r *. (1. -. cos (deg_to_rad 60.0))));
  printf "Area of sector  (central angle = 60 degrees) = %.2f\n"
    (60.0 /. 360.0 *. area);

  let p1 = point(0., 0.) in
  let p2 = point(0., -1.) in
  match circle_2pts_rad p1 p2 2., circle_2pts_rad p2 p1 2. with
  | Some p1, Some p2 ->
    printf "One of the center is (%.2f, %.2f)\n" p1.x p1.y;
    printf "The other center  is (%.2f, %.2f)\n" p2.x p2.y;
  | _ -> assert false

File: ch7/circles.py
import math

def DEG_to_RAD(d): return d*math.pi/180.0

def RAD_to_DEG(r): return r*180.0/math.pi

class point:
    def __init__(self, _x, _y):                   # int or double
        self.x = _x
        self.y = _y
    def getX(self):
        return self.x
    def getY(self):
        return self.y

# returns 0/1/2 for inside/border/outside, respectively
def insideCircle(p, c, r):                        # all integer version
    dx = p.getX()-c.getX()
    dy = p.getY()-c.getY()
    Euc = dx*dx + dy*dy
    rSq = r*r
    return 1 if Euc < rSq else (0 if Euc == rSq else -1)

def circle2PtsRad(p1, p2, r, c_list):
    # to get the other center, reverse p1 and p2
    d2 = (p1.getX()-p2.getX()) * (p1.getX()-p2.getX()) + (p1.getY()-p2.getY()) * (p1.getY()-p2.getY())
    det = r*r / d2 - 0.25
    if det < 0.0: return False
    h = math.sqrt(det)
    c_list[0].x = (p1.getX()+p2.getX()) * 0.5 + (p1.getY()-p2.getY()) * h
    c_list[0].y = (p1.getY()+p2.getY()) * 0.5 + (p2.getX()-p1.getX()) * h
    return True

def main():
    # circle equation, inside, border, outside
    pt = point(2, 2)
    r = 7
    inside = point(8, 2)
    print(insideCircle(inside, pt, r))            # 1, inside
    border = point (9, 2)
    print(insideCircle(border, pt, r))            # 0, at border
    outside = point(10, 2)
    print(insideCircle(outside, pt, r))           # -1, outside

    d = 2*r
    print("Diameter = ", "{:.2f}".format(d))
    c = math.pi*d
    print("Circumference (Perimeter) = ", "{:.2f}".format(c))
    A = math.pi*r*r
    print("Area of circle = ", "{:.2f}".format(A))

    print("Length of arc   (central angle = 60 degrees) = ", "{:.2f}".format(60.0/360.0 * c))
    print("Length of chord (central angle = 60 degrees) = ", "{:.2f}".format(math.sqrt((2*r*r) * (1 - math.cos(DEG_to_RAD(60.0))))))
    print("Area of sector  (central angle = 60 degrees) = ", "{:.2f}".format(60.0/360.0 * A))

    p1 = point(0, 0)
    p2 = point(0.0, -1.0)
    ans = [point(0, 0)]                           # use a wrapper
    circle2PtsRad(p1, p2, 2.0, ans) 
    print("One of the center is (", "{:.2f}".format(ans[0].getX()), ",", "{:.2f}".format(ans[0].getY()), ")")
    circle2PtsRad(p2, p1, 2.0, ans)               # reverse p1 with p2
    print("The other center is (", "{:.2f}".format(ans[0].getX()), ",", "{:.2f}".format(ans[0].getY()), ")")

main()
